pkg/pool-weighted/contracts/smart/LiquidityBootstrappingPool.sol

```diff
@@ -0,0 +1,501 @@
+
+    // The protocol fees will always be charged using the token associated with the max weight in the pool.
+    // Not worth packing - only referenced on join/exits
+    uint256 private _maxWeightTokenIndex;
+
+    // Storage for the current ongoing weight change
+    // Start weights are in _normalizedWeights
+    GradualUpdateParams private _gradualUpdateParams;
+
+    // Event declarations
+
+    event PublicSwapSet(bool isPublicSwap);
+    event GradualUpdateScheduled(uint256 startTime, uint256 endTime);
+
+    // Modifiers
+
+    /**
+     * @dev Reverts unless sender is the owner
+     */
+    modifier onlyOwner() {
+        _ensureOwner();
+        _;
+    }
+
+    constructor(
+        IVault vault,
+        string memory name,
+        string memory symbol,
+        IERC20[] memory tokens,
+        uint256[] memory normalizedWeights,
+        uint256 swapFeePercentage,
+        uint256 pauseWindowDuration,
+        uint256 bufferPeriodDuration,
+        address owner,
+        uint256 minDuration,
+        bool initialPublicSwap
+    )
+        BaseWeightedPool(
+            vault,
+            name,
+            symbol,
+            tokens,
+            new address[](tokens.length), // Pass the zero address: LBPs can't have asset managers
+            swapFeePercentage,
+            pauseWindowDuration,
+            bufferPeriodDuration,
+            owner
+        )
+    {
+        uint256 numTokens = tokens.length;
+        _require(numTokens <= _MAX_LBP_TOKENS, Errors.MAX_TOKENS);
+
+        InputHelpers.ensureInputLengthMatch(numTokens, normalizedWeights.length);
+
+        // Ensure each normalized weight is above the minimum and find the token index of the maximum weight
+        uint256 normalizedSum = 0;
+        uint256 maxWeightTokenIndex = 0;
+        uint256 maxNormalizedWeight = 0;
+        for (uint8 i = 0; i < numTokens; i++) {
+            uint256 normalizedWeight = normalizedWeights[i];
+            _require(normalizedWeight >= _MIN_WEIGHT, Errors.MIN_WEIGHT);
+
+            _setNormalizedWeight(normalizedWeight, i);
+
+            normalizedSum = normalizedSum.add(normalizedWeight);
+            if (normalizedWeight > maxNormalizedWeight) {
+                maxWeightTokenIndex = i;
+                maxNormalizedWeight = normalizedWeight;
+            }
+        }
+        // Ensure that the normalized weights sum to ONE
+        _require(normalizedSum == FixedPoint.ONE, Errors.NORMALIZED_WEIGHT_INVARIANT);
+
+        // Initial value; can change later if weights change
+        _maxWeightTokenIndex = maxWeightTokenIndex;
+
+        publicSwapEnabled = initialPublicSwap;
+        minWeightChangeDuration = minDuration;
+    }
+
+    // External functions
+
+    /**
+     * @dev Can pause/unpause trading
+     */
+    function setPublicSwap(bool publicSwap) external onlyOwner whenNotPaused nonReentrant {
+        publicSwapEnabled = publicSwap;
+
+        emit PublicSwapSet(publicSwap);
```

Comment:
Let's also emit this event in construction.

Reply:
Ok